sdk/src/program_error.rs

```diff
@@ -0,0 +1,127 @@
+use crate::instruction::InstructionError;
+use num_traits::ToPrimitive;
+
+/// Reasons the program may fail
+pub enum ProgramError {
+    /// CustomError allows programs to implement program-specific error types and see
+    /// them returned by the Solana runtime. A CustomError may be any type that is represented
+    /// as or serialized to a u32 integer.
+    ///
+    /// NOTE: u64 requires special serialization to avoid the loss of precision in JS clients and
+    /// so is not used for now.
+    CustomError(u32),
+    /// The arguments provided to a program instruction where invalid
+    InvalidArgument,
+    /// An instruction's data contents was invalid
+    InvalidInstructionData,
+    /// An account's data contents was invalid
+    InvalidAccountData,
+    /// An account's data was too small
+    AccountDataTooSmall,
+    /// An account's balance was too small to complete the instruction
+    InsufficientFunds,
+    /// The account did not have the expected program id
+    IncorrectProgramId,
+    /// A signature was required but not found
+    MissingRequiredSignature,
+    /// An initialize instruction was sent to an account that has already been initialized.
+    AccountAlreadyInitialized,
+    /// An attempt to operate on an account that hasn't been initialized.
+    UninitializedAccount,
+    /// The instruction expected additional account keys
+    NotEnoughAccountKeys,
+    /// Failed to borrow a reference to an account, already borrowed
+    AccountBorrowFailed,
+}
+
+/// 32bit representations of builtin program errors returned by the entry point
+const BUILTIN_ERROR_START: u32 = 0x8000_0000; // 31st bit set
+const INVALID_ARGUMENT: u32 = BUILTIN_ERROR_START + 1;
+const INVALID_INSTRUCTION_DATA: u32 = BUILTIN_ERROR_START + 2;
+const INVALID_ACCOUNT_DATA: u32 = BUILTIN_ERROR_START + 3;
+const ACCOUNT_DATA_TOO_SMALL: u32 = BUILTIN_ERROR_START + 4;
+const INSUFFICIENT_FUNDS: u32 = BUILTIN_ERROR_START + 5;
+const INCORRECT_PROGRAM_ID: u32 = BUILTIN_ERROR_START + 6;
+const MISSING_REQUIRED_SIGNATURES: u32 = BUILTIN_ERROR_START + 7;
+const ACCOUNT_ALREADY_INITIALIZED: u32 = BUILTIN_ERROR_START + 8;
+const UNINITIALIZED_ACCOUNT: u32 = BUILTIN_ERROR_START + 9;
+const NOT_ENOUGH_ACCOUNT_KEYS: u32 = BUILTIN_ERROR_START + 10;
+const ACCOUNT_BORROW_FAILED: u32 = BUILTIN_ERROR_START + 11;
+
+/// Is builtin if 31st bit is set
+fn is_builtin_error(error: u32) -> bool {
+    (error & BUILTIN_ERROR_START) != 0
+}
+
+/// If a program defined error conflicts with a builtin error
+/// its 30th bit is set before returning to distinguish it
+const CONFLICTING_ERROR: u32 = 0x4000_0000; // 30st bit set
+
+/// Is a conflicting error if 30th bit is set
+fn is_conflicting_error(error: u32) -> bool {
+    (error & CONFLICTING_ERROR) != 0
```

Comment:
error == 0 would be conflicting too right (it conflicts with SUCCESS)

Reply:
Not in this context, a "Conflicting error" is an error that has its 30th bit set.  An error that would conflict is 0 or any error that has its 31st bit set.  (checked and converted in the `entrypoint` macro) Those conflicting errors are turned into "Conflicting errors" by setting the 30th bit.  On the other side of the u32 return, the 30th bit is checked to see if this was a "conflicting error" and if so the bit is cleared and passed in `InstructionError::ConflictingError".  So if a user-defined an error value of 0 in their own custom enum it would result in an instruction error of `InstructionError::ConflictingError(0)`